### DISCUSSION TITLE
Query on variable of nested object

### DISCUSSION THREAD

**Discussion by vdkdamian** [netreconlab/Parse-Swift#32] [Q&A] [answered]
Back with another question.

I know it's possible to query relations. But I think I need something different.
I have a Pointer and I should need to be able to query all objects where a variable of that pointer is some specified value.

Eg.:

```
public struct Product: ParseObject {
    public var objectId: String?
    public var createdAt: Date?
    public var updatedAt: Date?
    public var ACL: ParseACL?
    public var originalData: Data?
    
    public var name: String?
    public var category: Category?
}
```

```
public struct Category: ParseObject {
    public var objectId: String?
    public var createdAt: Date?
    public var updatedAt: Date?
    public var ACL: ParseACL?
    public var originalData: Data?
    
    public var name: String?
}
```

I would need all Products where the name of Category for example == "Clothes".
I don't know if this is possible and how to do it.

Was thinking of something like this:

```
let productQuery = Product.query("category.name" == "Clothes")
        let products = try await productQuery.findAll()
```

**Comment by cbaker6** [comment]
This would be up to the server, not the Swift SDK. You would need to ask this question there to see if they allow such a query.

**Reply by cbaker6** [reply]
I'll note, I don't think a query like you are asking for needs to be supported on the server, you can achieve the same thing by using the correct combinations of `QueryContraint`, for example, [matchesKeyInQuery](https://netreconlab.github.io/Parse-Swift/release/documentation/parseswift/matcheskeyinquery(key:querykey:query:)), [doesNotMatchKeyInQuery](https://netreconlab.github.io/Parse-Swift/release/documentation/parseswift/doesnotmatchkeyinquery(key:querykey:query:)), and, or, nor, etc. You can see all of the possible `QueryConstraint`'s by looking at the `Functions` section of the [documentation](https://netreconlab.github.io/Parse-Swift/release/documentation/parseswift/)

**Reply by vdkdamian** [reply]
I'll take a look at it!

**Reply by vdkdamian** [reply]
Do you have some examples of use cases? 
I find it hard to see what it does.

**Reply by cbaker6** [reply]
You can search for `matchesKeyInQuery` in the [JS guide](https://docs.parseplatform.org/js/guide/#query-constraints) or the `matchesKey` [Objective-C guide](https://docs.parseplatform.org/ios/guide/#query-constraints). You can almost always find an equivalent in the JS guides

**Comment by jaysonng** [comment]
@vdkdamian check the inQuery section of [ParseQueryTests.swift](https://github.com/netreconlab/Parse-Swift/blob/3fd62f4831c0ecd964f1012e5cbf30526707999e/Tests/ParseSwiftTests/ParseQueryTests.swift#L1902) 

Basically I think what you need is to just get your Category object first and match that to the Product object

```
let inQuery = Category.query("name" == "Clothes")
let query = Product.query("category" == inQuery)
return try await query.findAll()
```

I think this code should get you what you need if I understood your original question correctly.

**Reply by vdkdamian** [reply]
I tried it, but it doesn't seem to work.

**Reply by jaysonng** [reply]
what's the error?

try  placing it in a do-catch

   ```
do {
            let inQuery = try Category.query("name" == "Clothes").includeAll()
            let query = Product.query("category" == inQuery).includeAll()
            return try await query.findAll()
        } catch  {
            throw error
        }
```

**Comment by cbaker6** [comment, marked as answer]
Your code will probably look something like below (the guides I posted in https://github.com/netreconlab/Parse-Swift/discussions/32#discussioncomment-4425264 further explain why):

```swift

let clothesQuery = Category.query("name" == "Clothes")
let productQuery = Product.query(matchesKeyInQuery(key: “category”,
                                                           queryKey: “name”,
                                                           query: clothesQuery))
do {
  let results = try await productQuery.find() // You should only use findAll when necessary is this isn’t an optimal query, you should only request what you need from the server
  print(results)
} catch {
  // Handle error
}
```

Note that if you don’t find something in the [Playgrounds](https://github.com/netreconlab/Parse-Swift/tree/main/ParseSwift.playground/Pages), it’s recommended to look through [SnapCat](https://github.com/netreconlab/SnapCat), [ParseCareKit](https://github.com/netreconlab/ParseCareKit) and [testcases](https://github.com/netreconlab/Parse-Swift/tree/main/Tests/ParseSwiftTests) as those are intended to demonstrate how to use the SDK.

**Reply by vdkdamian** [reply]
To everyone else needing this, try setting the queryKey to "objectId"

**Comment by vdkdamian** [comment]
Okay, so I was trying a lot of things and kept failing to get it working. 
I finally found how it works.
Apparently setting `queryKey` to "objectId", it worked for me.
Now I am thinking that that seems logically to me and would suggest a change in the code for simpler use.

What do you think if we change this function
```
public func matchesKeyInQuery <T>(key: String, queryKey: String, query: Query<T>) -> QueryConstraint {
    let select = QuerySelect(query: InQuery(query: query), key: queryKey)
    return QueryConstraint(key: key, value: select, comparator: .select)
}
```

to this function:
```
public func matchesKeyInQuery <T>(key: String, queryKey: String = "objectId", query: Query<T>) -> QueryConstraint {
    let select = QuerySelect(query: InQuery(query: query), key: queryKey)
    return QueryConstraint(key: key, value: select, comparator: .select)
}
```

To me it seems like most people would need the basic use and set the queryKey to "objectId". But let me know what you think!

**Reply by cbaker6** [reply]
From the guides I posted and using this in other apps with the Swift and Objective-C SDK (some examples below). This change isn’t necessary, there’s probably something else going on with your code, how you have your Parse Classes configured, or server.

https://github.com/netreconlab/SnapCat/blob/a296aab39e83ec7ee5a13a488911f34c83b1e1cb/SnapCat/Activity/ActivityViewModel.swift#L16-L20

https://github.com/netreconlab/SnapCat/blob/2490ec4026d2bef89e79adc45ee9d1d41ccf99fc/SnapCat/Home/TimeLineViewModel.swift#L46-L53

**Reply by vdkdamian** [reply]
Okay, thanks for the help tho!

**Reply by vdkdamian** [reply]
I was trying examples based on SnapCat, but couldn't get them working.
Could it be it has something to do with the fact that I am using customObjectId's? 
Can't think of anything else that's different on my configuration.

**Reply by cbaker6** [reply]
It seems what you need is a simpler version that shouldn't need `matchesKeyInQuery`. From what you've listed, @jaysonng's example of using [==](https://netreconlab.github.io/Parse-Swift/release/documentation/parseswift/==(_:_:)-998hz) in https://github.com/netreconlab/Parse-Swift/discussions/32#discussioncomment-4440742 should have worked for you if `category` really is a `ParseObject`: 

```swift
let inQuery = Category.query("name" == "Clothes")
let query = Product.query("category" == inQuery)
do {
  let results = try await query.find()
  print(results)
} catch {
  // Handle error
}
```

**Reply by vdkdamian** [reply]
Ah yes, I checked and it's working indeed. Must have done something wrong when trying @jaysonng example.

**Reply by jaysonng** [reply]
I was wondering why my code didn't work for you as it works fine on my code. I don't have a matchesKeyInQuery in my project (but started looking into it and playing with it since this discussion began).

**Reply by vdkdamian** [reply]
@jaysonng Think I did something wrong with my first query. But it's working now
